### ISSUE TITLE
allowed_roles may not work?

### ISSUE THREAD

**Issue by lenscas** [serenity-rs/serenity#941]
For a project I'm currently working I wanted to use the allowed_roles to put a configuration command behind the "Moderator" or the "ThankBotManager" role. However the line `#[allowed_roles("Moderator", "ThankBotManager")]` doesn't seem to do anything as even without roles I can get access to this command?

I also let someone else test it in case it had something to do with me being the creator of the server but in both cases the other person could also see the command in `!help` and was able to execute it.

Maybe I am doing something wrong? The exact line can be found at https://github.com/lenscas/thanks_bot/blob/master/src/commands/config.rs#L10 

For now I worked around it by also checking in the command itself for those roles so this bug is not a blocker for me.
I'm currently targeting version `"0.9.0-rc.0"`

If there is more information that that I can provide to help fix it then I'm more than happy to, just tell me what you need and how to get it :)

Side note: so far the library is very nice to use. So, thanks for making it :)

**Comment by arqunis** [comment]
> Maybe I am doing something wrong?

You are not. This is entirely the fault of the framework. `allowed_roles` is a defined option for groups and commands alongside documentation describing its function. However, it's actually unimplemented. The framework never checks whether the user has the roles when parsing a command invocation or even after that. [The only place where it *does* check is in the help commands.](https://github.com/serenity-rs/serenity/blob/current/src/framework/standard/help_commands.rs#L291). But again, the function isn't used anywhere. It's defined, but left abandoned.

**Comment by arqunis** [comment]
I was wrong. [The framework *does* check the roles of the user](https://github.com/serenity-rs/serenity/blob/current/src/framework/standard/parse/mod.rs#L270). It is likely the problem occured because the framework ignored the `LackingRole` error when parsing commands. I recommend testing [this commit](https://github.com/serenity-rs/serenity/commit/323875f5a9ca9b22202e6e291bfca0c78794cfdb) to see if that has fixed your issue as well.

Here's how to do it. `Cargo.toml`:
```toml
[dependencies.serenity]
git = "https://github.com/serenity-rs/serenity"
rev = "323875f5a9ca9b22202e6e291bfca0c78794cfdb"
```

**Comment by lenscas** [comment]
Sorry for the late reply.

Just tested it and that seems to work both for preventing the command to show up in !help AND by preventing the command from being run.

Thanks for looking into it and fixing it for me :)

**Comment by arqunis** [comment]
> fixing it for me :)

It was a mere coincidence that your problem got fixed as well. I didn't know what the cause of it was, so I didn't know where to start. @bdashore3 had issues with the `DispatchError::UnrecognisedCommand` error being returned even when `DispatchError::LackingPermissions` or `DispatchError::LackingRole` were returned from a command, but ignored. In the case of `UnrecognisedCommand` I knew exactly where I had to make changes to fix its behaviour. 

Anyway, I'm glad it works for you. I'm closing this issue.
